### PR TITLE
Dev 99 103 api refactor

### PR DESCRIPTION
**Pull request by donaldcampbelljr**
Adds `select_records`, `retrieve_one`, `retrieve_many ` interface, cursor_pagination for DBBackend.

Also refactored to store all configuration attributes in `psm.cfg`.

Changed magic functions, `__setitem__ `and `__getitem__`, to act as a wrapper for `report` and `retrieve` respectively.

Address #99 and #103

- [x] Delete `dynamic_filter` function and other, as we replaced it with new function

### CODE REVIEW

File: pipestat/backends/abstract.py [hunk 1]
```diff
@@ -215,3 +215,15 @@ def remove_record(
...
+
+def select_records(
+    self,
+    columns: Optional[List[str]] = None,
+    filter_conditions: Optional[List[Tuple[str, str, Union[str, List[str]]]]] = None,
+    json_filter_conditions: Optional[List[Tuple[str, str, str]]] = None,
+    limit: Optional[int] = 1000,
+    cursor: Optional[int] = None,
+    bool_operator: Optional[str] = "AND",
+) -> List[Any]:
```
Comment (khoroshevskyi) [inline pipestat/backends/abstract.py:228]:
```suggestion
    self,
    columns: Optional[List[str]] = None,
    filter_conditions: Optional[List[Dict[str, Any]]] = None,
    limit: Optional[int] = 1000,
    cursor: Optional[int] = None,
    bool_operator: Optional[str] = "AND",
) -> Dict[str, Any]:
```

File: pipestat/backends/db_backend/dbbackend.py [hunk 2]
```diff
@@ -622,6 +622,71 @@ def select(
...
+        filter_conditions: Optional[List[Tuple[str, str, Union[str, List[str]]]]] = None,
+        limit: Optional[int] = 1000,
+        cursor: Optional[int] = None,
+        bool_operator: Optional[str] = "AND",
+    ) -> List[Any]:
+        """
+        Perform a `SELECT` on the table
+
+        :param list[str] columns: columns to include in the result
+        :param list[tuple(str,str,str)] [(key,operator,value)] filter_conditions: e.g. [("id", "eq", 1)], operator list:
```
Comment (khoroshevskyi) [inline pipestat/backends/db_backend/dbbackend.py:637]:
```suggestion
        :param list[dict]  filter_conditions: e.g. [{"key": ["id"], "operator": "eq", "value": 1)], operator list:
```

File: pipestat/pipestat.py [hunk 3]
```diff
@@ -118,56 +118,70 @@ def __init__(
...
+        validate(self.cfg[CONFIG_KEY].exp, cfg_schema)
 
-        self[SCHEMA_PATH] = self[CONFIG_KEY].priority_get(
+        self.cfg[SCHEMA_PATH] = self.cfg[CONFIG_KEY].priority_get(
             "schema_path", env_var=ENV_VARS["schema"], override=schema_path
         )  # schema_path if schema_path is not None else None
         self.process_schema(schema_path)
 
-        self[RECORD_IDENTIFIER] = record_identifier
+        # self[RECORD_IDENTIFIER] = record_identifier
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:146]: Can you please clean this. to many comments are. And probably noone will use them

File: pipestat/pipestat.py [hunk 4]
```diff
@@ -217,47 +231,53 @@ def __str__(self):
         return res
 
     def __getitem__(self, key):
-        return self.store[self._keytransform(key)]
+        # This is a wrapper for the retrieve function:
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:234]: I like It!

File: pipestat/pipestat.py [hunk 5]
```diff
@@ -217,47 +231,53 @@ def __str__(self):
...
+        result = self.retrieve(record_identifier=key)
+        return result
 
     def __setitem__(self, key, value):
-        self.store[self._keytransform(key)] = value
+        # This is a wrapper for the report function:
+        result = self.report(record_identifier=key, values=value)
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:240]: I like it too!

File: pipestat/pipestat.py [hunk 6]
```diff
@@ -217,47 +231,53 @@ def __str__(self):
...
+        return result
 
     def __delitem__(self, key):
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:243]: you can call `pipestat.remove` in this magic method

File: pipestat/pipestat.py [hunk 7]
```diff
@@ -537,6 +558,81 @@ def retrieve_distinct(
             result = self.backend.select_distinct(columns=columns)
         return result
 
+    @require_backend
+    def select_records(
+        self,
+        columns: Optional[List[str]] = None,
+        filter_conditions: Optional[List[Tuple[str, str, Union[str, List[str]]]]] = None,
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:565]:
```suggestion
        filter_conditions: Optional[List[Dict[str, Any]]] = None,
```

File: pipestat/pipestat.py [hunk 8]
```diff
@@ -537,6 +558,81 @@ def retrieve_distinct(
...
+        return self.select_records(filter_conditions=filter_conditions)
+
+    def retrieve_many(
+        self,
+        record_identifiers: List[str],
+    ) -> Union[Any, Dict[str, Any]]:
+        """
+        Retrieve multiple records given a list of strings
+
+        returns list of records
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:619]: add ":param record_identifiers ..."

File: pipestat/pipestat.py [hunk 9]
```diff
@@ -537,6 +558,81 @@ def retrieve_distinct(
...
+            filter = {
+                "key": "record_identifier",
+                "operator": "eq",
+                "value": r_id,
+            }
+
+            filter_conditions.append(filter)
+
+        return self.select_records(filter_conditions=filter_conditions, bool_operator="OR")
+
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:635]:
```suggestion
        filter = {
            "key": "record_identifier",
            "operator": "in",
            "value": record_identifiers,
        }

        return self.select_records(filter_conditions=[filter])

```

File: pipestat/pipestat.py [hunk 10]
```diff
@@ -574,7 +670,7 @@ def retrieve(
                     record_identifier, result_identifier, limit, offset
                 )
 
-        r_id = record_identifier or self.record_identifier
+        r_id = record_identifier or self.cfg[RECORD_IDENTIFIER]
```
Comment (khoroshevskyi) [inline pipestat/pipestat.py:673]: Nice!

File: tests/data/sample_output_schema.yaml [hunk 11]
```diff
@@ -22,3 +22,14 @@ samples:
...
+#    type: object
+#    description: "This is an example file"
+#    properties:
+#      prop1:
+#        description: "This is an example file"
+#        type: object
+#        properties:
+#          prop2:
+#            description: "This is an example file"
+#            type: integer
```
Comment (khoroshevskyi) [inline tests/data/sample_output_schema.yaml:35]:
```suggestion
```

**Review by khoroshevskyi — CHANGES_REQUESTED** [review]
Looks good, few changes needed

File: pipestat/backends/db_backend/db_helpers.py
Comment (khoroshevskyi) [inline pipestat/backends/db_backend/db_helpers.py:1]: Delete dynamic filter, as it is replaced by selection filter

**Review by khoroshevskyi — CHANGES_REQUESTED** [review]
One critical change needed

File: pipestat/backends/db_backend/db_helpers.py
Comment (donaldcampbelljr) [reply pipestat/backends/db_backend/db_helpers.py:1]: I left _all_ of the old functions in pipestat for now until it was decided if we want to keep them for backwards compat or not. Dynamic filter is required for the original `select` function.

**Review by khoroshevskyi — APPROVED** [review]
LGTM

**Comment by donaldcampbelljr** [comment]
- [ ] remove unused funcs (e.g. select_txt)
- [ ] make wrapper around list_recent_results (but use `select_records` backend)
- [ ] use select_records for filebackend as well.

**Comment by donaldcampbelljr** [comment]
I added the select_records function to the filebackend, but it is missing the ability to filter by result_identifiers (columns). Currently it only retrieves the entire record.

**Comment by donaldcampbelljr** [comment]
- [ ] get_one_record, get_records, retrieve still exist on the backends, we should be able to remove them and simply use `select_records`.

**Comment by donaldcampbelljr** [comment]
removing retrieve is still a work in progress. also need to decide where to place `RecordNotFoundError` since it lists in Retrieve but not select_records.

**Comment by donaldcampbelljr** [comment]
retrieve is now completely removed.

I attempted to re-implement RecordNotFound for retrieve_one and retrieve_many but its not quite behaving as expected.

**Pull request merged**